src/os/bluestore/BlueStore_DB_Hash.cc

```diff
@@ -0,0 +1,678 @@
+#include "BlueStore.h"
+#include "kv/RocksDBStore.h"
+#include "include/ceph_hash.h"
+#define dout_context cct
+#define dout_subsys ceph_subsys_bluestore
+#undef dout_prefix
+#define dout_prefix *_dout << "shard-db "
+
+
+
+
+
+
+
+class BlueStore_DB_Hash : public KeyValueDB {
+public:
+  typedef std::map<std::string, size_t> ShardingSchema;
+private:
+  RocksDBStore* db;
+  CephContext* cct;
+  const rocksdb::Comparator* comparator;
+  ShardingSchema sharding_schema;
+  typedef std::map<std::string, std::vector<KeyValueDB::ColumnFamilyHandle> > ActiveShards;
+  ActiveShards shards;
+
+public:
+  BlueStore_DB_Hash(RocksDBStore* db, const ShardingSchema& sharding_schema)
+  : db(db), cct(db->cct), sharding_schema(sharding_schema) {
+    comparator = db->rocksdb_options.comparator;
+    ceph_assert(db);
+  }
+  virtual ~BlueStore_DB_Hash() {
+    delete db;
+  }
+private:
+  int open_shards() {
+    for (auto& s_it: sharding_schema) {
+      for (size_t i = 0; i < s_it.second; i++) {
+        std::string name = s_it.first + "-" + to_string(i);
+        auto cf_handle = db->column_family_handle(name);
+        dout(0) << "Column family '" << name << "' handle: " << (void*)cf_handle.priv << " " << dendl;
+        shards[s_it.first].push_back(cf_handle);
+      }
+    }
+    return 0;
+  }
+  KeyValueDB::ColumnFamilyHandle get_db_shard(const std::string &prefix, const char *k, size_t keylen) {
+    auto it = shards.find(prefix);
+    ceph_assert(it != shards.end());
+    unsigned hash = ceph_str_hash_linux(k, keylen);
+    return it->second[hash % it->second.size()];
```

Comment:
I wonder if we have an opportunity to do something smarter, here.  What if, for each prefix, we define a custom hash function.  For OMAP keys, for instance, the key starts with the %16x omap head. If we hashed *just* that part of the string, we could keep all keys for a single object in the same shard.

Maybe that's a bad example if we want huge omap objects to span shards (do we??).

A better one is probably the onode key space, where we could keep the onode and all of its shards in the same CF/shard by using just the first portion of the key that has the pool id and object hash position....

Reply:
A simple key prefix length might be enough, actually.. where we just cap keylen at some prefix length that varies per prefix based on how they key names are structured.

Reply:
This could also avoid some of the overhead of rechecking for a valid sort in the next() on the iterator, because we could avoid that check if the prefix portion of the key hasn't changed.

Reply:
I understand value of such change. I will add it.